src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,688 @@
+        /// of the method invocations where a parameter was added
+        /// </summary>
+        private static SyntaxNode GenerateNewInvocationExpressionForTrampoline(ISyntaxFactsService syntaxFacts,
+            SyntaxEditor editor, SyntaxNode currentInvocation, SyntaxNode invocationExpression,
+            string newMethodIdentifier, int insertionIndex)
+        {
+            var invocationArguments = syntaxFacts.GetArgumentsOfInvocationExpression(currentInvocation);
+            var methodName = editor.Generator.IdentifierName(newMethodIdentifier);
+            var expressionFromInvocation = syntaxFacts.GetExpressionOfInvocationExpression(invocationExpression);
+            var newMethodInvocation = editor.Generator.InvocationExpression(methodName, invocationArguments);
+            var allArguments = invocationArguments.Insert(insertionIndex, newMethodInvocation);
+            return editor.Generator.InvocationExpression(expressionFromInvocation, allArguments);
+        }
+
+        /// <summary>
+        /// Goes through all of the invocations and replaces the expression with identifiers from the invocation 
+        /// arguments and rewrites the call site with the updated expression as a new argument
+        /// </summary>
+        private async Task<Solution> RewriteSolutionWithNewMethodAsync(
+            SemanticDocument semanticDocument, TExpressionSyntax expression, IMethodSymbol methodSymbol,
+            bool allOccurrences, string parameterName,
+            ImmutableDictionary<Document, List<TInvocationExpressionSyntax>> callSites, bool trampoline,
+            CancellationToken cancellationToken)
+        {
+            var firstCallSite = callSites.Keys.First();
+            var modifiedSolution = firstCallSite.Project.Solution;
+            var newMethodIdentifier = methodSymbol.Name + "_" + parameterName;
+
+            foreach (var grouping in callSites.GroupBy(kvp => kvp.Key.Project))
+            {
+                var project = grouping.Key;
+                var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+                foreach (var (document, invocationExpressionList) in grouping)
+                {
+                    var generator = SyntaxGenerator.GetGenerator(document);
+                    var semanticFacts = document.GetRequiredLanguageService<ISemanticFactsService>();
+                    var invocationSemanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                    var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                    var editor = new SyntaxEditor(root, generator);
+                    var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+                    var oldMethodDeclaration = expression.FirstAncestorOrSelf<SyntaxNode>(node => IsContainedInParameterizedDeclaration(node), ascendOutOfTrivia: true)!;
```

Comment:
i don't know why you are computing this for every document you hit.

Reply:
the oldMethodDeclaration? I am moving that to the very top pf the project. The root and editor need to be recomputed for every document, right? SyntaxFacts I understand can be also be moved to the top of the method